src/main/java/org/jboss/dmr/stream/ModelReader.java

```diff
@@ -0,0 +1,196 @@
+     * @return string the parsing cursor is pointing to
+     * @exception IllegalStateException if cursor isn't pointing to DMR String
+     */
+    String getString();
+
+    /**
+     * Returns <code>true</code> if current DMR parsing event is DMR <code>expression</code>, <code>false</code> otherwise.
+     * Users have to call {@link #next()} before calling this method.
+     * @return true if the parsing cursor position points to DMR expression, false otherwise
+     */
+    boolean isExpression();
+
+    /**
+     * Converts available context data to <code>expression</code>.
+     * Users have to call {@link #next()} and should call {@link #isExpression()} before calling this method.
+     * @return expression the parsing cursor is pointing to
+     * @exception IllegalStateException if cursor isn't pointing to DMR expression
+     */
+    String getExpression();
+
+    /**
+     * Returns <code>true</code> if current DMR parsing event is DMR <code>bytes</code>, <code>false</code> otherwise.
+     * Users have to call {@link #next()} before calling this method.
+     * @return true if the parsing cursor position points to DMR bytes, false otherwise
+     */
+    boolean isBytes();
+
+    /**
+     * Converts available context data to <code>bytes</code>.
+     * Users have to call {@link #next()} and should call {@link #isBytes()} before calling this method.
+     * @return bytes the parsing cursor is pointing to
+     * @exception IllegalStateException if cursor isn't pointing to DMR Bytes
+     */
+    byte[] getBytes();
+
+    /**
+     * Returns <code>true</code> if current DMR parsing event is DMR <code>boolean</code> token, <code>false</code> otherwise.
+     * Users have to call {@link #next()} before calling this method.
+     * @return true if the parsing cursor points to DMR boolean tokens, false otherwise
+     */
+    boolean isBoolean();
+
+    /**
+     * Converts available context data to <code>boolean</code>.
+     * Users have to call {@link #next()} and should call {@link #isBoolean()} before calling this method.
+     * @return boolean value the parsing cursor is pointing to
+     * @exception IllegalStateException if cursor isn't pointing to DMR boolean token
+     */
+    boolean getBoolean();
+
+    /**
+     * Returns <code>true</code> if current DMR parsing event is DMR <code>number</code>, <code>false</code> otherwise.
+     * Users have to call {@link #next()} before calling this method.
+     * @return true if the parsing cursor points to DMR number, false otherwise
+     */
+    boolean isNumber();
+
+    /**
+     * Converts available context data to <code>Number</code> instance.
+     * Users have to call {@link #next()} and should call {@link #isNumber()} before calling this method.
+     * @return number value the parsing cursor is pointing to
+     * @exception NumberFormatException if DMR number is not convertible to <code>number</code>
+     */
+    Number getNumber();
```

Comment:
I don't like generalizing this.  It's better to have separate getInt(), getLong() which do not box and getBigInteger() which returns BigInteger.